tests/integration/upgrades/test_upgrades.py

```diff
@@ -0,0 +1,58 @@
+#!/usr/bin/env python3
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import pytest
+from pytest_operator.plugin import OpsTest
+
+from ..ha_tests.helpers import find_unit
+from ..helpers import (
+    APP_NAME,
+    check_or_scale_app,
+    get_app_name,
+    get_password,
+    set_password,
+)
+
+
+@pytest.mark.skip("Missing upgrade code for now")
+@pytest.mark.group(1)
+@pytest.mark.abort_on_fail
+async def test_build_and_deploy(ops_test: OpsTest):
+    app_name = await get_app_name(ops_test)
+
+    if app_name:
+        await check_or_scale_app(ops_test, app_name, required_units=3)
+        return
+
+    app_name = APP_NAME
+
+    await ops_test.model.deploy(
+        app_name,
+        application_name=app_name,
+        num_units=3,
+        series="jammy",
+        channel="6/edge",
+    )
+    await ops_test.model.wait_for_idle(
+        apps=[app_name], status="active", timeout=1000, idle_period=120
+    )
+
+
+@pytest.mark.skip("Missing upgrade code for now")
+@pytest.mark.group(1)
+@pytest.mark.abort_on_fail
+async def test_upgrade_password_change_fail(ops_test: OpsTest):
+    app_name = await get_app_name(ops_test)
+    leader_id = await find_unit(ops_test, leader=True, app_name=app_name)
+
+    current_password = await get_password(ops_test, leader_id, app_name=app_name)
+    new_charm = await ops_test.build_charm(".")
+    await ops_test.model.applications[app_name].refresh(path=new_charm)
+    results = await set_password(ops_test, leader_id, password="0xdeadbeef", app_name=app_name)
```

Comment:
just FYI I am not sure if we will be able to guarantee that an upgrade is running *while* running the action `set_password`

the reason is, is that if the snap revision doesn't change then the UX is simply `juju refresh` and the refresh happens instantaneously (or very quickly for that matter) 

if the snap revision *does* change then the UX is `juju refresh; juju resume-upgrade` and in between those two actions we can guarantee that an upgrade is running and we could do  `juju refresh; juju run mongodb-k8s/leader set-password; juju resume-upgrade`

So I am not sure actually if it makes sense to have this here or not. But I would love to hear your thoughts and expertise

Reply:
That's exactly what I was wondering.
I added this because it was part of the DoD of [the ticket](https://warthogs.atlassian.net/browse/DPE-5372).
If we go to upgrade V3, it will be easier to do that because we can set `pause-after-refresh` to `all`.
I did this (fully skipped) as a first version to be updated when we would have upgrades V3.